lib/mini_racer.rb

```diff
@@ -340,19 +341,37 @@ def stop_attached
     end
 
     def timeout(&blk)
-      return blk.call unless @timeout
+      return blk.call unless @timeout or @cputime_limit
 
       mutex = Mutex.new
       done = false
 
       rp,wp = IO.pipe
 
+      start_monotonic_time = Process.clock_gettime Process::CLOCK_MONOTONIC, :millisecond
+      thread_clock = nil
+      if @cputime_limit then
+        # this call might fail, so condition on actually needing it
+        thread_clock = current_thread_clock
+        start_cpu_time = Process.clock_gettime thread_clock, :millisecond
```

Comment:
More bad news-- the first param of Process.clock_gettime calls itself a 'clock_id' but they have made a really crappy decision here to name their ENUM the same thing as the underlying method's ID:

https://ruby-doc.org/core-2.5.6/Process.html#method-c-clock_gettime

Which caused me a bit of confusion, and is definitely not what I would call a best practice

Reply:
I was also not able to find an overload method that allows for calling the function with an actual ID

Reply:
After a chat with GPT I have received some answers and ideas:

https://chatgpt.com/share/67055d40-47b4-8002-8318-a5d28ab53b91

We will need to pack the whole thing up into an extension method that we call from this watchdog thread, but can just use the thread id instead I think.

Reply:
> More bad news-- the first param of Process.clock_gettime calls itself a 'clock_id' but they have made a really crappy decision here to name their ENUM the same thing as the underlying method's ID:
> 
> https://ruby-doc.org/core-2.5.6/Process.html#method-c-clock_gettime
> 
> Which caused me a bit of confusion, and is definitely not what I would call a best practice

> I was also not able to find an overload method that allows for calling the function with an actual ID

Ruby does a really funny thing here where *on supported platforms*, the `Process::CLOCK_MONOTONIC` &c constants are literally just the underlying POSIX API values, meaning that as long as you're on one of those you can just stick whatever ID you get from anywhere else into `Process.clock_gettime`. Otherwise those constants are marker values that they then use to do translation when you call `Process.clock_gettime` with them.

It's a whole mess, because they're potentially trying to emulate an API that doesn't exist on the target system; but... yeah, it "just works" on supported platforms.

Reply:
> After a chat with GPT I have received some answers and ideas:
> 
> https://chatgpt.com/share/67055d40-47b4-8002-8318-a5d28ab53b91
> 
> We will need to pack the whole thing up into an extension method that we call from this watchdog thread, but can just use the thread id instead I think.

This sounds promising but I'd like to see `man`page excerpts or... *any* sort of documentation. Apple does have *some* of it online but... uh... it's not very good. [cf. `thread_info` """docs"""](https://developer.apple.com/documentation/kernel/1418630-thread_info)

Reply:
> > More bad news-- the first param of Process.clock_gettime calls itself a 'clock_id' but they have made a really crappy decision here to name their ENUM the same thing as the underlying method's ID:
> > https://ruby-doc.org/core-2.5.6/Process.html#method-c-clock_gettime
> > Which caused me a bit of confusion, and is definitely not what I would call a best practice
> 
> > I was also not able to find an overload method that allows for calling the function with an actual ID
> 
> Ruby does a really funny thing here where _on supported platforms_, the `Process::CLOCK_MONOTONIC` &c constants are literally just the underlying POSIX API values, meaning that as long as you're on one of those you can just stick whatever ID you get from anywhere else into `Process.clock_gettime`. Otherwise those constants are marker values that they then use to do translation when you call `Process.clock_gettime` with them.
> 
> It's a whole mess, because they're potentially trying to emulate an API that doesn't exist on the target system; but... yeah, it "just works" on supported platforms.

Wild.

As described in other comment, I'm going to be forking our main even harder. (removing mac as platform) now that I have a dev env.